table/block_based_table_reader.cc

```diff
@@ -2551,12 +2549,29 @@ void BlockBasedTableIterator<TBlockIter, TValue>::FindKeyBackward() {
   // code simplicity.
 }
 
+template <class TBlockIter, typename TValue>
+void BlockBasedTableIterator<TBlockIter, TValue>::CheckOutOfBound() {
+  if (read_options_.iterate_upper_bound != nullptr &&
+      block_iter_points_to_real_block_ && block_iter_.Valid()) {
+    Slice current_user_key = block_iter_.key();
+    if (key_includes_seq_) {
+      current_user_key = ExtractUserKey(current_user_key);
+    }
```

Comment:
This is a little bit trickier than I thought. BlockBasedTableIterator also covers the second level of partitioned index. While I can't think of anything wrong with the code, I may be a good idea for @maysamyabandeh to verify it too.

Reply:
@maysamyabandeh mind take a quick look? Thanks!

Reply:
@yiwu-arbug I did but the look was not exactly quick ;) I think it would have helped if we had this in the PR description:
> To optimize checking for bounds, we need comparing the bounds with index key as well. However BlockBasedTableIterator doesn't know whether its index iterator is internally using user keys or internal keys. The patch fixes that by exposing this detail of the index iterator to the BlockBasedTableIterator class.

Reply:
done.